README
Quickstart
==========

$ ./configure
$ make

Optional (you can try camsource without installing it):
# make install

Take a look at the included sample config file (camsource.conf.example).
It should be somewhat usable for experimental uses (although the
hardware section <camdev> may need tweaking), but you should change it
to suit your needs if you plan to use camsource. Put the config file in
a place where camsource can find it (such as .camsource in your home
dir, or /usr/local/etc/camsource.conf). The start camsource:

$ camsource

Or, if you didn't install it:

$ src/camsource

If camsource starts up without errors, it will go into background
automatically. Note that any errors from the grabbing thread show up
_after_ the logfile has been started, so you should check the logfile
for any errors (by default .camsource.log in your home dir). If
everything worked right and camsource shows up in your process list
(multiple time as it's multithreaded), feel free to try it
out. With the default config, point your browser to
<http://localhost:9192/> to see a 320x240 snapshot. The "/large" path
gives you an unresized snapshot. If the images appear mirrored, edit
your config and disable the "flip" module (it's enabled by default to
mirror all images).

With the default config, you have multiple options for getting an
animated webcam display:

http://host:9192/
http://host:9192/large
  Using a javascript html page, you can continually reload these images
  to get an animation. The "camserv" software includes such a javascript
  snippet (javascript.txt), see <http://cserv.sourceforge.net/>.

http://host:9192/multipart
http://host:9192/multipart-large
  These urls produce a multipart jpeg stream at a given fps. Netscape
  and mozilla can display and animate those without plugin (even though
  you may need to embed the url as image into a html page). This jpeg
  stream is compatible to what "camserv" produces (see above for url).
  To support multipart jpegs on other browsers such as ie, a java applet
  is required. Cambozola does an excellent job at this. See
  <http://www.charliemouse.com/code/cambozola/index.html>.

http://host:8888/
http://host:8889/
  These are pseudo-urls and not accessible through a browser. They
  produce an output compatible to "webcam_server," and the java applet
  included with webcam_server can be use to access these urls and
  animate the images. Note that even though webcam_server allows you to
  get a snapshot image when you point a browser at its url, the
  camsource module doesn't do this; point your browser at one of the
  other urls above instead. You can find webcam_server and the java
  applet at <http://donn.dyndns.org/portalofnnod/webcam_server.html>.

http://host:9192/largequal
  This is the unresized grabbed frame with a high jpeg compression
  quality (80). It is about twice the size of the default "/large"
  image. Use for your personal viewing pleasure or something.



Details (What is Camsource anyway?)
===================================

Camsource opens a video4linux device, grabs frames from it and does
various things with those frames. What it does with them depends solely
on the config file, which is in xml format.

Camsource uses plugins (also called modules) to provide functionality.
The plugin configuration happens through the xml config file. The core
camsource program, as well as the plugins, are heavily multithreaded, to
make sure different plugins don't interfere with each other, and with
the grabbing process.

Currently included plugins are:

wc_serv
  Produces output compatible with webcam_server. You can use its java
  applet to view the images.

http
  Generic http module, designed to answer requests by web browsers. It
  lets you define several virtual paths, each with distinct properties.
  Currently it is able to produce a single still image, as well as an
  animage multipart jpeg stream at a given framerate, usable on
  netscape/mozilla browsers or using a java applet such as Cambozola.
  Note that the paths you define can also include a query string such as
  "/image.jpg?large"; the requested url will always be matched exactly
  to the letter.

filewrite
  Periodically saves a snapshot image to a local jpeg file. 'Nuff said.

ftpup
  Works similar to the filewrite module, only that it uploads the snapshot
  to an ftp server instead of writing it out locally.

jpeg_comp
  This plugin doesn't do anything by itself, but provides jpeg
  compression functionality for various other modules. It will be
  auto-loaded as dependency, but you can give a default jpeg compression
  quality level for this module in the config file.

flip (filter)
  Lets you flip an image horizontally or vertically before doing
  anything else with it. Useful for cameras which produce mirrored
  images (such as my creative webcam 5).

resize (filter)
  Lets you scale an image to a specific size. Actually it doesn't
  "scale" much, it just does nearest-neighbor matching (read: bad
  quality). The filter lets you specify the target image size as a fixed
  width/height value pair, or as a percentage of the original image.

text (filter)
  This filter can add a text overlay containing a timestamp to an
  image. It has some special functionality with regards to the config
  file, which allow it to have a global config, and then override
  certain options in the per-module filter definition. See the
  example config file for details.
  The text to display can be set in the config file, or it can be
  read on demand from a file or from a command's stdout (pipe).

rgbbgr (filter)
  Swaps the red and blue information. Activate it in case your camera
  produces images with this reversed.

rotate (filter)
  This filter can rotate the image left or right by 90 degrees (or in
  other words, by 90 or 270 degrees). Rotating by 180 degrees is not
  supported, as the same result can be achieved by "flip"ping the image
  both horizontally and vertically. Keep in mind that after rotating,
  the image's width and height values will be swapped, so when you
  intend to use the rotate filter frequently together with the resize
  module, you should probably specify the final image size as a
  percentage (scale) instead of a fixed height/width pair.

bw (filter)
  Converts a color image into artistic grayscale / black & white.



The config file should be pretty self explanatory. Every possible
feature is present in the default config. Filters can be applied
globally (that is, to every grabbed frame), or on a per-module basis.
Note that the module is responsible for calling the filters, this
doesn't happen automatically. In the case of the http module, filters
can be applied to every frame handled by the http module, as well as on
a per-vpath basis.

Camsource looks in several places for its config file. First it looks
for it as ".camsource" in your home dir, then checks for a
camsource.conf file in /etc, /usr/local/etc and in the local dir. It
also looks for a "camsource" file in /etc/defaults and
/usr/local/etc/defaults. You can also specify a custom config file to
use on the command line.

The command line switch "-c" has been created to dump the video4linux
video capability information to console, then exit. Use this if you're
unsure about which frame sizes your camera supports. Optionally you can
specify another device file to open and query (defaults to /dev/video0).



Links
=====

Homepage: http://camsource.sf.net/


$Id: README,v 1.14 2002/10/02 16:43:09 dfx Exp $



ChangeLog
$Id: ChangeLog,v 1.56 2003/05/01 10:35:39 dfx Exp $

May 1 2003
	- Released 0.7.0.

Apr 20 2003
	- Brought back the -c flag with changed behavior to
	  fit the new modular input plugin design.
	- Applied (modified) patch sent by Nigel Roberts
	  implementing the <autobrightness> feature. Credits
	  also go to Stefan Meyer for the original code from
	  wwwcam.

Mar 9 2003
	- Http requests can now take a query string, which will
	  be ignored when looking for a matching config vpath.
	- Fixed ftp uploading for ftpds which give multi-line
	  replies in a certain format.
	- Support basic http authentication.

Feb 15 2003
	- Added experimental xwd (x11 screen shot) input plugin.

Jan 26 2003
	- Modularized the video input code, removed camdev.* files.
	  Their functionality was moved to libinput_v4l. The
	  '-c' command line switch doesn't work now. Will fix.
	- Libfilewrite and libftpup now have the capability of
	  timestamping filenames.

Jan 23 2003
	- Applied patch by Jaco Greeff to give libftpup a better
	  sleep time precision, compensating for long upload times.

Jan 20 2003
	- Changed the '~' filename postfix to '.tmp' in libfilewrite
	  and libftpup.

Jan 18 2003
	- Added scripts/camsource2bmp.pl example script.
	- Added the smallest filter ever: color inverting.
	- Implemented -r and -s/-k command line switches to restart
	  and shut down camsource.

Jan 16 2003
	- Reworked _private handling a bit. Avoid repeated parsing
	  or large xml trees.
	- Simplyfied per-instance filter storage to make use of
	  libxml's _private struct member.

Jan 15 2003
	- Added <cmd> option to camdev section to support external
	  trigger for capture on/off events.

Jan 14 2003
	- The http module can now server raw image data, plus it now
	  supports 'connection: keep-alive'. Combining those two
	  features, it is possible for third-party processes to
	  access the video data.
	- Fixed old jpeg compression crash bug
	- Added capability of handling multiple input v4l devices
	  at the same time.

Jan 5 2003
	- Added motiondetect custom config. Filter works better
	  now too.
	- Added regrab filter.
	- Added per-instance context storage for filters.
	- Modified a few filters to make use of the new per-
	  instance storage space to save a bit cpu (?)

Jan 4 2003
	- Added a preliminary motiondetect filter. More to do.
	- Changed the filter handling a bit, to make it possble
	  to have filters which "swallow" the image.
	- Added vloopback module, courtesy of Cyril Rocheteau.
	- Tried to make up some login to autodetect whether overlay
	  is supported, and gracefully ignore it if it isn't. Same
	  goes for unsupported fps settings (e.g. bttv).

Dec 31 2002
	- Included patch sent by Cyril Rocheteau to allow camsource
	  to work with drivers which don't support overlay
	  capturing (e.g. meye).

--- 0.6.0 ---

Dec 17 2002
	- Released 0.6.0.

Dec 13 2002
	- Added norm and picture metric video device settings.

Nov 21 2002
	- Added support for the mmap v4l api. Fixes bttv support.

--- 0.5.1 ---

Nov 14 2002
	- Released 0.5.1.
	- Changed palette ordering so yuv420p gets priority over
	  yuv420 due to ambiguity in the specs.

--- 0.5.0 ---

Oct 7 2002
	- Released 0.5.0.

Oct  2 2002
	- Added bw filter

Sep 28 2002
	- Supports more palettes. They're untested and probably broken
	- Added more information to palette struct and added palette probing
	  to -c switch output.
	- Added experimental grayscale palette support
	- Pumped version to 0.5.0a.

--- 0.4.0 ---

Sep 27 2002
	- Released camsource 0.4.0.

Sep 26 2002
	- Fixed various compilation issues, including compatibility with
	  libxml1.

Sep 25 2002
	- Added version information to modules

Sep 24 2002
	- Added <scale> option to resize filter
	- Added a rotate image filter
	- Per request, added feature to filewrite and ftpup module to allow
	  grabbing of a single frame, then exit.
	- Fixed small logging typo
	- Pumped version to 0.4.0a.

--- 0.3.0 ---

Sep 22 2002
	- Released 0.3.0
	- Fixed small typo bug in http logging

Sep 21 2002
	- Removed unnecessary headers from make install
	- libtext can now read text from a file and from a pipe/command.
	- Added quick and dirty rgb <-> bgr module.

Sep 20 2002
	- With nonblocking sockets, the delay before closing a sending
	  socket shouldn't be necessary any more.
	- Added text module to all worker modules in conf file, plus docs
	- Finished work on libtext with lots of fancy stuff for now
	- Started work in text overlay module
	- Made all (?) socket stuff time out properly and added better
	  error/condition reporting
	- Converted all socket stuff to nonblocking sockets
	- Made custom jpeg compression quality levels possible (<jpegqual>)
	- Pumped devel version number to 0.2.1

--- 0.2.0 ---

Sep 19 2002
	- Released 0.2.0
	- Added timeout option to socket_accept()
	- Finished ftpup functionality with adding safemode
	- Lots of code cleanups and improvements in ftpup

Sep 18 2002
	- Started work on ftpup module
	- Added a few socket functions
	- Added <cmd> tag to filewrite mod
	- Added config_homedir() for "~/" expansion
	- Added grabbed image aging and struct grab_ctx handling to prevent
	  that an outdated image gets output accidently
	- Improved logging facilities
	- Converted lots of functions to static to prevent global symbol /
	  namespace clashes

Sep 17 2002
	- Created config option to select input channel
	- Sanity check grabbing width/height values
	- Can now specify "default", "max" etc in certain config options
	- Video device is opened before going to background, so errors show up
	  on console.
	- Pumped development version number to 0.1.1
	- Added filewrite module

--- 0.1.0 ---

Sep 16 2002
	- Released version 0.1.0



src/http.c
#include <stdio.h>
#include <errno.h>
#include <string.h>
#include <unistd.h>
#include <stdlib.h>
#include <libxml/parser.h>

#include "config.h"

#include "http.h"
#define MODULE_THREAD
#include "module.h"
#include "xmlhelp.h"
#include "socket.h"
#include "grab.h"
#include "filter.h"
#include "jpeg.h"
#include "mod_handle.h"
#include "log.h"
#include "image.h"

#define MODNAME "http"

struct vpathcfg {
	int fps;
	char *auth;
	int raw:1;
};

static int http_load_conf(struct http_ctx *, xmlNodePtr);
static void *http_conn(void *);
static int http_path_ismatch(xmlNodePtr, char *);
static void http_err(struct peer *, char *, char *);
static void http_fill_vpathcfg(xmlNodePtr, struct vpathcfg *);
static void unbase64(unsigned char *, unsigned char *);

char *name = MODNAME;
char *version = VERSION;
char *deps[] =
{
	"jpeg_comp",
	"socket",
	NULL
};

int
init(struct module_ctx *mod_ctx)
{
	int ret;
	struct http_ctx *http_ctx;
	
	if (!mod_ctx->node)
		return -1;
	
	http_ctx = malloc(sizeof(*http_ctx));
	ret = http_load_conf(http_ctx, mod_ctx->node);
	if (ret)
	{
		free(http_ctx);
		return -1;
	}
	
	http_ctx->listen_fd = socket_listen(http_ctx->port, 0);
	if (http_ctx->listen_fd == -1)
	{
		log_log(MODNAME, "Failed to open listen socket: %s\n", strerror(errno));
		free(http_ctx);
		return -1;
	}
	
	mod_ctx->custom = http_ctx;
	
	return 0;
}

void *
thread(void *mod_ctx)
{
	int ret;
	struct http_ctx *http_ctx;
	struct http_peer *http_peer;
	
	http_ctx = ((struct module_ctx *) mod_ctx)->custom;
	
	for (;;)
	{
		http_peer = malloc(sizeof(*http_peer));
		http_peer->mod_ctx = mod_ctx;
		ret = socket_accept_thread(http_ctx->listen_fd, &http_peer->peer, http_conn, http_peer);
		if (ret)
		{
			free(http_peer);
			log_log(MODNAME, "accept() error: %s\n", strerror(errno));
			sleep(1);
			continue;
		}
	}
	
	return NULL;
}

static
int
http_load_conf(struct http_ctx *ctx, xmlNodePtr node)
{
	memset(ctx, 0, sizeof(*ctx));
	
	ctx->listen_fd = -1;
	ctx->port = 9192;
	
	for (node = node->xml_children; node; node = node->next)
	{
		if (xml_isnode(node, "port"))
			ctx->port = xml_atoi(node, ctx->port);
	}
	
	if (ctx->port <= 0 || ctx->port > 0xffff)
	{
		log_log(MODNAME, "Invalid port %u\n", ctx->port);
		return -1;
	}
	
	return 0;
}

static
void *
http_conn(void *peer_p)
{
	struct http_peer http_peer;
	char buf[1024];
	int ret;
	char *url;
	char *p;
	char *httpver;
	xmlNodePtr subnode;
	struct image img;
	struct grab_ctx idx;
	struct jpegbuf jpegbuf;
	int count;
	char *error;
	struct vpathcfg vpathcfg;
	char *contenttype;
	char addheaders[256];
	char authorization[256];
	
	memcpy(&http_peer, peer_p, sizeof(http_peer));
	free(peer_p);
	
	log_log(MODNAME, "New connection from %s:%i\n",
		socket_ip(&http_peer.peer), socket_port(&http_peer.peer));
	
	count = 0;
	
	do {
		ret = socket_readline(&http_peer.peer, buf, sizeof(buf), 20000);
		if (ret)
		{
readlineerr:
			switch (ret)
			{
			case -1:
				log_log(MODNAME, "Read error on connection from %s:%i (%s)\n",
					socket_ip(&http_peer.peer), socket_port(&http_peer.peer), strerror(errno));
				break;
			case -2:
				log_log(MODNAME, "Timeout on connection from %s:%i\n",
					socket_ip(&http_peer.peer), socket_port(&http_peer.peer));
				break;
			case -3:
				if (count)
					break;
				log_log(MODNAME, "EOF on connection from %s:%i before full HTTP request was received\n",
					socket_ip(&http_peer.peer), socket_port(&http_peer.peer));
				break;
			}
			goto closenout;
		}
		
		if (strncmp("GET ", buf, 4))
		{
			log_log(MODNAME, "Invalid HTTP request (not GET) from %s:%i\n",
				socket_ip(&http_peer.peer), socket_port(&http_peer.peer));
			goto closenout;
		}
		
		url = buf + 4;
		
		p = strchr(url, ' ');
		if (!p)
			httpver = NULL;
		else
		{
			*p++ = '\0';
			if (strncmp("HTTP/", p, 5))
				httpver = NULL;
			else
			{
				httpver = p + 5;
				if (!*httpver)
					httpver = NULL;
			}
		}
		
		if (!*url || !httpver)
		{
			log_log(MODNAME, "Invalid HTTP request or version from %s:%i\n",
				socket_ip(&http_peer.peer), socket_port(&http_peer.peer));
			goto closenout;
		}
		
		p = strchr(url, '?');
		if (p)
			*p = '\0';
		
		log_log(MODNAME, "Request for %s from %s:%i\n",
			url, socket_ip(&http_peer.peer), socket_port(&http_peer.peer));
	
		error = NULL;
		for (subnode = http_peer.mod_ctx->node->xml_children; subnode; subnode = subnode->next)
		{
			if (!xml_isnode(subnode, "vpath"))
				continue;
			if (!http_path_ismatch(subnode, url))
				continue;
			goto match;
		}
		error = "404 Not found";
		
match:
		*authorization = '\0';
		for (;;)
		{
			ret = socket_readline(&http_peer.peer, buf, sizeof(buf), 20000);
			if (ret)
				goto readlineerr;
			if (!*buf)
				break;
			if (!strncasecmp(buf, "authorization:", 14)) {
				p = buf + 14;
				while (*p == ' ')
					p++;
				if (strncasecmp(p, "basic ", 6))
					continue;
				p += 6;
				while (*p == ' ')
					p++;
				if (strlen(p) >= sizeof(authorization)) {
					log_log(MODNAME, "Warning: very long Authorization: header received\n");
					continue;
				}
				unbase64(authorization, p);
			}
		}
		
		if (error) {
			http_err(&http_peer.peer, error, NULL);
			goto closenout;
		}
		
		http_fill_vpathcfg(subnode, &vpathcfg);
		if (vpathcfg.auth) {
			if (strcmp(authorization, vpathcfg.auth)) {
				http_err(&http_peer.peer, "401 Unauthorized", "WWW-Authenticate: Basic realm=\"camsource\"\r\n");
				goto closenout;
			}
		}

		if (vpathcfg.fps > 0)
		{
			snprintf(buf, sizeof(buf) - 1,
				"HTTP/1.0 200 OK\r\n"
				"Server: " PACKAGE_STRING "\r\n"
				"Connection: close\r\n"		/* <-- does that make sense? */
				"Pragma: no-cache\r\n"
				"Cache-Control: no-cache\r\n"
				"Content-Type: multipart/x-mixed-replace;boundary=Juigt9G7bb7sfdgsdZGIDFsjhn\r\n"
				"\r\n"
				"--Juigt9G7bb7sfdgsdZGIDFsjhn\r\n");
			socket_write(&http_peer.peer, buf, strlen(buf), 20000);
		}
		
		memset(&idx, 0, sizeof(idx));
		do
		{
			filter_get_image(&img, &idx, http_peer.mod_ctx->node, subnode, NULL);
			if (!vpathcfg.raw) {
				jpeg_compress(&jpegbuf, &img, subnode);
				contenttype = "image/jpeg";
				addheaders[0] = '\0';
			}
			else {
				jpegbuf.buf = img.buf;
				jpegbuf.bufsize = img.bufsize;
				contenttype = "application/octet-stream";
				snprintf(addheaders, sizeof(addheaders) - 1,
					"X-Image-Width: %i\r\n"
					"X-Image-Height: %i\r\n",
					img.x, img.y);
			}
			
			if (!vpathcfg.fps)
				snprintf(buf, sizeof(buf) - 1,
					"HTTP/1.0 200 OK\r\n"
					"Server: " PACKAGE_STRING "\r\n"
					"Content-Length: %i\r\n"
					"Connection: Keep-alive\r\n"
					"Pragma: no-cache\r\n"
					"Cache-Control: no-cache\r\n"
					"Content-Type: %s\r\n"
					"%s"
					"\r\n",
					jpegbuf.bufsize, contenttype,
					addheaders);
			else
				snprintf(buf, sizeof(buf) - 1,
					"Content-Length: %i\r\n"
					"Content-Type: %s\r\n"
					"%s"
					"\r\n",
					jpegbuf.bufsize, contenttype,
					addheaders);
			ret = socket_write(&http_peer.peer, buf, strlen(buf), 20000);
			if (ret > 0)
				ret = socket_write(&http_peer.peer, jpegbuf.buf, jpegbuf.bufsize, 20000);
	
			count++;
	
			if (vpathcfg.fps && ret > 0)
			{
				ret = socket_write(&http_peer.peer, "\r\n--Juigt9G7bb7sfdgsdZGIDFsjhn\r\n", 32, 20000);
				usleep(1000000 / vpathcfg.fps);
			}
			
			image_destroy(&img);
			if (!vpathcfg.raw)
				free(jpegbuf.buf);
		}
		while (vpathcfg.fps > 0 && ret > 0);
	}
	while (ret > 0);

closenout:
	log_log(MODNAME, "Closing connection from %s:%i, %i frame(s) served\n",
		socket_ip(&http_peer.peer), socket_port(&http_peer.peer),
		count);

	socket_close(&http_peer.peer);
	return NULL;
}

static
int
http_path_ismatch(xmlNodePtr node, char *path)
{
	int ret = 0;
	char *active;
	
	active = xml_attrval(node, "active");
	if (active && strcmp(active, "yes"))
		return 0;
	
	for (node = node->xml_children; node; node = node->next)
	{
		if (xml_isnode(node, "disable")) {
			if (!strcmp("yes", xml_getcontent_def(node, "")))
				return 0;
		}
		
		if (!xml_isnode(node, "path"))
			continue;
		if (!strcmp(path, xml_getcontent_def(node, "")))
			ret = 1;
	}
	
	return ret;
}

static
void
http_err(struct peer *peer, char *err, char *addheaders)
{
	char buf[1024];
	
	snprintf(buf, sizeof(buf) - 1,
		"HTTP/1.0 %s\r\n"
		"Server: " PACKAGE_STRING "\r\n"
		"Content-Length: %i\r\n"
		"Connection: close\r\n"
		"Pragma: no-cache\r\n"
		"Cache-Control: no-cache\r\n"
		"Content-Type: text/html\r\n"
		"%s"
		"\r\n"
		"<html><body>%s</body></html>\r\n",
		err, strlen(err) + 28, addheaders ? : "", err);
	socket_write(peer, buf, strlen(buf), 20000);
}

static
void
http_fill_vpathcfg(xmlNodePtr node, struct vpathcfg *cfg)
{
	memset(cfg, 0, sizeof(*cfg));
	
	for (node = node->xml_children; node; node = node->next)
	{
		if (xml_isnode(node, "fps"))
			cfg->fps = xml_atoi(node, 0);
		else if (xml_isnode(node, "raw") && !strcmp("yes", xml_getcontent_def(node, "")))
			cfg->raw = 1;
		else if (xml_isnode(node, "auth"))
			cfg->auth = xml_getcontent(node);
	}
}

static void
unbase64(unsigned char *dst, unsigned char *src)
{
	/* yuck */
	static unsigned char lookup[256] = {
		/*   0 */ 255,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
		/*  16 */ 0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
		/*  32 */ 0,0,0,0,0,0,0,0,0,0,0,62,0,0,0,63,
		/*  48 */ 52,53,54,55,56,57,58,59,60,61,0,0,0,255,0,0,
		/*  64 */ 0,0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,
		/*  80 */ 15,16,17,18,19,20,21,22,23,24,25,0,0,0,0,0,
		/*  96 */ 0,26,27,28,30,30,31,32,33,34,35,36,37,38,39,40,
		/* 128 */ 41,42,43,44,45,46,47,48,49,50,51,0,0,0,0,0,
	};
	
	int extract[4];
	
	while (*src) {
		if ((extract[0] = lookup[src[0]]) == 255 || (extract[1] = lookup[src[1]]) == 255)
			break;
		
		*dst++ = (extract[0] << 2) | (extract[1] >> 4);
		
		if ((extract[2] = lookup[src[2]]) == 255)
			break;
			
		*dst++ = (extract[1] << 4) | (extract[2] >> 2);

		if ((extract[3] = lookup[src[3]]) == 255)
			break;
			
		*dst++ = (extract[2] << 6) | extract[3];
		
		src += 4;
	}
	
	*dst = '\0';
}

